bliss/models/fnp.py

```diff
@@ -0,0 +1,475 @@
+        input_list = [yR_encoded]
+        if self.use_x:
+            input_list.append(XR.unsqueeze(0))
+
+        ## If we look at differences in U values, we need to increase the dimension
+        ## (each representative member looks different to each dependent member)
+        if self.use_u_diff:
+            u_diff = u.unsqueeze(1) - uR.unsqueeze(0)
+            for i, x in enumerate(input_list):
+                input_list[i] = x.unsqueeze(1)
+            input_list.append(u_diff.unsqueeze(0))
+
+        rep_R = self.f(self.cat(*input_list))
+        return rep_R
+
+
+# **************************
+# Representation Poolers
+# **************************
+class AveragePooler(nn.Module):
+    """
+    Pools together representations by taking a sum.
+    """
+
+    def __init__(
+        self,
+        dim_z,
+    ):
+        super().__init__()
+        self.dim_z = dim_z
+
+        # normalizes the graph such that inner products correspond to averages of the parents
+        self.norm_graph = lambda x: x / (torch.sum(x, 1, keepdim=True) + 1e-8)
+
+    def forward(self, rep_R, GA):
+        W = self.norm_graph(GA)
+        pz_all = torch.matmul(W, rep_R)
+        return pz_all
+
+
+class SetPooler(nn.Module):
+    """
+    Pools together representations using a set transformer architecture.
+    See https://arxiv.org/abs/1810.00825.
+    """
+
+    def __init__(
+        self,
+        dim_rep_R,
+        dim_z,
+        pooling_layers,
+        set_transformer,
+        st_numheads,
+    ):
+        super().__init__()
+        self.dim_rep_R = dim_rep_R
+        self.dim_z = dim_z
+        self.pooling_layers = pooling_layers
+        self.set_transformer = set_transformer
+        self.st_numheads = st_numheads
+
+        # normalizes the graph such that inner products correspond to averages of the parents
+        self.norm_graph = lambda x: x / (torch.sum(x, 1, keepdim=True) + 1e-8)
+
+        if not self.set_transformer:
+            self.pool_net = MLP(self.dim_rep_R, self.pooling_layers, 2 * self.dim_z)
+            self.settrans = None
+        else:
+            self.pool_net = None
+            dim_in = self.dim_rep_R
+            sabs = [SAB(dim_in, dim_in, nh) for nh in self.st_numheads]
+            self.settrans = nn.Sequential(
+                *sabs,
+                PMA(dim_in, 2, 1, squeeze_out=True),
+                MLP(dim_in, self.pooling_layers, 2 * self.dim_z),
+            )
+
+    def forward(self, rep_R, GA):
+        if not self.set_transformer:
+            rep_pooled = GA.unsqueeze(0).unsqueeze(-1).mul(rep_R).sum(2)
+            pz_all = self.pool_net(rep_pooled)
+        else:
+            pz_all = self.settrans(GA.unsqueeze(0).unsqueeze(-1).mul(rep_R))
+        return pz_all
+
+
+# *********************
+# Set Transformer
+# *********************
+# This implementation is taken from https://github.com/juho-lee/set_transformer/,
+# which comes with the MIT license.
+
+
+class MAB(nn.Module):
```

Comment:
Can you use [this](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) instead of implementing from scratch? Or are you doing something sufficiently different?

Reply:
I moved this code over from the set_transformer repo mostly intact; maybe the MultiheadAttention got added to torch later. 

I would like to use the built-in torch module eventually, but I need to study it a bit more to make sure it is a drop-in replacement. I can open a separate issue after merging

Reply:
Sounds good, just wanted you to be aware this existed. My sense is that it's pretty new but not completely sure. Thanks!